p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	encodedLocalFingerprint, err := encodeDTLSFingerprint(listenerFingerprint)
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	certMultiaddress, err := ma.NewMultiaddr(fmt.Sprintf("/webrtc/certhash/%s", encodedLocalFingerprint))
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerMultiaddr = listenerMultiaddr.Encapsulate(certMultiaddress)
+
+	return newListener(
+		t,
+		listenerMultiaddr,
+		socket,
+		t.privKey,
+		t.webrtcConfig,
+	)
+}
+
+func (t *WebRTCTransport) Dial(
+	ctx context.Context,
+	remoteMultiaddr ma.Multiaddr,
+	p peer.ID,
+) (tpt.CapableConn, error) {
+	remoteMultihash, err := decodeRemoteFingerprint(remoteMultiaddr)
+	if err != nil {
+		log.Debugf("could not decode remote multiaddr: %v", err)
+		return nil, err
+	}
+
+	rnw, rhost, err := manet.DialArgs(remoteMultiaddr)
+	if err != nil {
+		log.Debugf("could not generate dialargs: %v", err)
+		return nil, err
+	}
+
+	raddr, err := net.ResolveUDPAddr(rnw, rhost)
+	if err != nil {
+		log.Debugf("could not resolve udp address: %v", err)
+		return nil, err
+	}
+
+	localFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		log.Debugf("could not get local fingerprint: %v", err)
+		return nil, err
+	}
+
+	// @mxinden instead of encoding the local fingerprint we
+	// instead generate a random uuid as the connection ufrag.
+	// The only requirement here is that the ufrag and password
+	// must be equal, which will allow the server to determine
+	// the password using the STUN message.
+	ufrag := uuid.New().String()
+
+	se := webrtc.SettingEngine{}
+	se.DetachDataChannels()
+	se.SetICECredentials(ufrag, ufrag)
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(se))
+
+	pc, err := api.NewPeerConnection(t.webrtcConfig)
+	if err != nil {
+		log.Debugf("could not instantiate peerconnection: %v", err)
+		return nil, err
+	}
+
+	// We need to set negotiated = true for this channel on both
+	// the client and server to avoid DCEP errors.
+	dc, err := pc.CreateDataChannel("data", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(1),
+	})
+
+	if err != nil {
+		_ = pc.Close()
+		log.Debugf("could not create datachannel: %v", err)
+		return nil, err
+	}
+
+	opened := make(chan *DataChannel, 1)
+	errChan := make(chan error, 1)
+	dc.OnOpen(func() {
+		detached, err := dc.Detach()
+		if err != nil {
+			log.Debugf("could not detach datachannel: %v", err)
+			errChan <- err
+		}
+		cp, err := dc.Transport().Transport().ICETransport().GetSelectedCandidatePair()
+		if err != nil {
+			log.Debugf("could not fetch selected candidate pair: %v", err)
+			errChan <- err
+			return
+		}
+
+		laddr := &net.UDPAddr{IP: net.ParseIP(cp.Local.Address), Port: int(cp.Local.Port)}
```

Comment:
we handle the error case when it would return nil. This is only nil when the ICE agent itself is nil https://github.com/pion/webrtc/blob/v3.1.43/icetransport.go#L47